src/type/__tests__/validation-test.js

```diff
@@ -687,6 +687,80 @@ describe('Type System: Input Objects must have fields', () => {
     ]);
   });
 
+  it('accepts an Input Object with breakable circular reference', () => {
```

Comment:
I like how you considered all the possibilities of breaking a circular reference in the first test case, to ensure this does not cause false-positives.